# Tools

## Data preprocessing

Load a month of monthly Neighborhood Patterns, then prepare data for IPF.

In [1]:
# ! conda config --prepend channels conda-forge
# ! conda  install  --strict-channel-priority geopandas  

In [36]:
%load_ext autoreload
%autoreload 2
import os
import random
import sqlite3 
import numpy as np
import json
import math
from tqdm.notebook import tqdm
from tqdm import tqdm
tqdm.pandas()

from ipfn import ipfn

import pandas as pd
# import geopandas as gpd
import matplotlib.pyplot as plt

import Advan_operator as ad_op  

pd.set_option('display.max_columns', None)

import logging
# Create a logger
logger_name = 'all_logger'
logger = logging.getLogger(logger_name)


The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


### Load monthly Neighborhood Patterns

#### Asign file locations

In [37]:
# data_dir = r'F:\SafeGraph\Advan_2023_API\Neighborhood_Patterns'   # Lenovo


# data_dir = r'K:\SafeGraph\Advan_2023_API\Neighborhood_Patterns'   # Dell 
# save_dir = r'D:\OneDrive_PSU\OneDrive - The Pennsylvania State University\Research_doc\Wild_fire\hourly_map_test'  # Dell



# ACS_file = r"D:\OneDrive_PSU\OneDrive - The Pennsylvania State University\Research_data\Safegraph_bias\cbg_acs_2019_county_tract_new20230929_cleaned.csv"
ACS_file = r"D:\OneDrive_PSU\OneDrive - The Pennsylvania State University\Research_data\Safegraph_bias\cbg_acs_2019_county_tract_new20230929_cleaned.csv"

home_panel_fname = f"D:\OneDrive_PSU\OneDrive - The Pennsylvania State University\Research_doc\Wild_fire\Monthly_Patterns_home_panel_summary_2019_2023.zip"
# home_panel_fname = r"E:\OneDrive_PSU\OneDrive - The Pennsylvania State University\Research_doc\Wild_fire\home_panel_summary_2019_2023_restated_20240119.csv.gz"  # large change!!!
# https://community.deweydata.io/t/residing-device-count-in-2023-07-and-2023-08-surged-abnomorally/26675/15

# data_dir = r'F:\SafeGraph\Advan_2023_API\Neighborhood_Patterns'   # Lenovo 
# save_dir = r'D:\OneDrive_PSU\OneDrive - The Pennsylvania State University\Research_doc\Wild_fire\hourly_map_test'  # Lenovo
# ACS_file = r"F:\Research\Wild_fire\cbg_acs_2019_county_tract_new20230929_cleaned.csv"
# home_panel_fname = f"F:\SafeGraph\Advan_2023_API\Monthly_Patterns_home_panel_summary\Monthly_Patterns_home_panel_summary_2019_2023.csv"

# 2018 desktop
# save_dir = r'E:\OneDrive_PSU\OneDrive - The Pennsylvania State University\Research_doc\Wild_fire\hourly_map_test'  # 2018 desktop
save_dir = r'E:\OneDrive_PSU\OneDrive - The Pennsylvania State University\Research_doc\Wild_fire\hourly_map_test_2024_home_panel'  # 2018 desktop

# data_dir = r'D:\SafeGraph\Advan_2023_API\Neighborhood_Patterns'  
data_dir = r'D:\SafeGraph\Advan_2024_API\Neighborhood_Patterns'    

ACS_file = r"E:\OneDrive_PSU\OneDrive - The Pennsylvania State University\Research_data\Safegraph_bias\cbg_acs_2019_county_tract_new20230929_cleaned.csv"
# home_panel_fname = f"E:\OneDrive_PSU\OneDrive - The Pennsylvania State University\Research_doc\Wild_fire\Monthly_Patterns_home_panel_summary_2019_2023.zip"
home_panel_fname = r"E:\OneDrive_PSU\OneDrive - The Pennsylvania State University\Research_doc\Wild_fire\home_panel_summary_2019_2023_restated_20240119.csv.gz"  ## large change!!!
# https://community.deweydata.io/t/residing-device-count-in-2023-07-and-2023-08-surged-abnomorally/26675/15


year = '2023'
month = 6

os.makedirs(save_dir, exist_ok=True)




month_dir = os.path.join(data_dir, str(year), f'{month:02}', '01')
no_dict_colum_fname = os.path.join(save_dir, f"NP_dict_column_{year}_{month:02}.csv")
device_home_areas_fname = os.path.join(save_dir, f"NP_split_device_home_areas_{year}_{month:02}.csv")
adjusted_dwelling_time_stop_CBG_fname     = os.path.join(save_dir, f"adjusted_dwelling_time_stop_CBG_{year}_{month:02}.csv")
adjusted_dwelling_time_stop_national_fname = os.path.join(save_dir, f"adjusted_dwelling_time_stop_national_{year}_{month:02}.csv")
person_per_device_fname = os.path.join(save_dir, f"person_per_device_{year}_{month:02}.csv")
restored_CBG_hourly_vistor_count_fname = os.path.join(save_dir, f"restored_hourly_vistor_count_CBG_{year}_{month:02}.csv")
restored_target_hourly_vistor_count_fname = os.path.join(save_dir, f"restored_hourly_vistor_count_target_{year}_{month:02}.csv")
CBG_device_stop_destination_distribution_fname  = os.path.join(save_dir, f"CBG_device_stop_destination_distribution_{year}_{month:02}.csv")
CBG_device_stop_origin_distribution_fname = os.path.join(save_dir, f"CBG_device_stop_origin_distribution_{year}_{month:02}.csv")
IPF_results_fname = os.path.join(save_dir, f"IPF_origin_outgoer_hourly_{year}{month:02}.csv")
CBG_hourly_population_fname = os.path.join(save_dir, f"CBG_population_hourly_{year}{month:02}.csv")

CBG_origin_outbound_distribution_sampling_rate_fname = os.path.join(save_dir, f"CBG_origin_outbound_distribution_sampling_rate_{year}{month:02}.csv")
CBG_destination_inbound_per_device_fname = os.path.join(save_dir, f"CBG_destination_inbound_per_device_{year}{month:02}.csv")

CBG_destination_inbound_fname = os.path.join(save_dir, f"CBG_destination_inbound_{year}{month:02}.csv")
inbound_hourly_df_fname = os.path.join(save_dir, f"inbound_hourly_{year}{month:02}.csv")
CBG_agregated_inbound_fname =  os.path.join(save_dir, f"CBG_agregated_inbound_{year}{month:02}.csv")
hourly_agregated_inbound_fname =  os.path.join(save_dir, f"hourly_agregated_inbound_{year}{month:02}.csv")

 
# month_dir = r'K:\SafeGraph\Advan_2023_API\Neighborhood_Patterns\2023\06\01'
# CBG_list_fname = r"D:\OneDrive_PSU\OneDrive - The Pennsylvania State University\Research_doc\Wild_fire\ACS_2019_CBG_list.csv"

sqlite_fname = r'F:\Research\Wild_fire\Neighborhood_202306.db'


print("Month directory:", month_dir)
print("Save_dir directory:", save_dir)
print("ACS_file:", ACS_file)
print("home_panel file name:", home_panel_fname)

print("No dict colum file name:", no_dict_colum_fname)
print("Device home areas file name:", device_home_areas_fname)
print("Device person_per_device file name:", person_per_device_fname)


print("Device adjusted_dwelling_time_stop_CBG_fname:", adjusted_dwelling_time_stop_CBG_fname)
print("Device adjusted_dwelling_time_stop_national_fname:", adjusted_dwelling_time_stop_national_fname)

print("Restored_hourly_vistor_count_CBG file name:   ", restored_CBG_hourly_vistor_count_fname)
print("Restored_hourly_vistor_count_target file name:", restored_target_hourly_vistor_count_fname)

print("CBG_device_stop_destination_distribution file name:   ", CBG_device_stop_destination_distribution_fname)
print("CBG_device_stop_origin_distribution file name:", CBG_device_stop_origin_distribution_fname)

print("IPF_results file name:", IPF_results_fname)
print("CBG hourly population file name:", CBG_hourly_population_fname)

print("CBG_origin_outbound_distribution_sampling_rate file name:", CBG_origin_outbound_distribution_sampling_rate_fname)


Month directory: D:\SafeGraph\Advan_2024_API\Neighborhood_Patterns\2023\06\01
Save_dir directory: E:\OneDrive_PSU\OneDrive - The Pennsylvania State University\Research_doc\Wild_fire\hourly_map_test_2024_home_panel
ACS_file: E:\OneDrive_PSU\OneDrive - The Pennsylvania State University\Research_data\Safegraph_bias\cbg_acs_2019_county_tract_new20230929_cleaned.csv
home_panel file name: E:\OneDrive_PSU\OneDrive - The Pennsylvania State University\Research_doc\Wild_fire\home_panel_summary_2019_2023_restated_20240119.csv.gz
No dict colum file name: E:\OneDrive_PSU\OneDrive - The Pennsylvania State University\Research_doc\Wild_fire\hourly_map_test_2024_home_panel\NP_dict_column_2023_06.csv
Device home areas file name: E:\OneDrive_PSU\OneDrive - The Pennsylvania State University\Research_doc\Wild_fire\hourly_map_test_2024_home_panel\NP_split_device_home_areas_2023_06.csv
Device person_per_device file name: E:\OneDrive_PSU\OneDrive - The Pennsylvania State University\Research_doc\Wild_fire\hour

## Load monthly neighborhood patterns

In [38]:
use_cols =  ['AREA', 
             "DEVICE_HOME_AREAS",
            "STOPS_BY_DAY", 
            "RAW_STOP_COUNTS", 
            "RAW_DEVICE_COUNTS", 
            "MEDIAN_DWELL", 
            "STOPS_BY_EACH_HOUR"]
month_df = ad_op.load_neighborhood_monthly_folder(folder=month_dir, extions=['gz'], use_cols=use_cols, verbose=True)
month_df = month_df.sort_values("AREA")
month_df['AREA'] = month_df['AREA'].astype(str).str.zfill(12)
month_df = month_df[use_cols]
month_df

AREA                                  DEVICE_HOME_AREAS  \
17609  010010201001  {"010010201001":84,"010010208021":63,"01051031...   
3175   010010201002  {"010010201002":132,"010010209002":21,"0100102...   
17751  010010202001  {"010010208012":71,"010010202001":71,"01001020...   
13849  010010202002  {"010010205003":193,"010010205002":193,"010010...   
10035  010010203001  {"010010203001":528,"010010208024":111,"010010...   
...             ...                                                ...   
19948  780309611001  {"780309611001":39,"780309608001":21,"78030961...   
20862  780309611002  {"780309611002":48,"780309611001":44,"78030960...   
17718  780309612001  {"780309611001":21,"780309608001":20,"78030960...   
3191   780309612002  {"780309611001":28,"780309611002":23,"78030960...   
17524  780309900000  {"780309608001":11,"780309601003":6,"780309900...   

                                            STOPS_BY_DAY  RAW_STOP_COUNTS  \
17609  [286,254,196,74,246,284,287,227,272,212,249,20...             7936   
3175   [248,206,132,132,191,155,156,98,94,172,118,130...             5227   
17751  [229,61,189,267,93,72,191,112,215,177,52,152,5...             5516   
13849  [1097,900,666,861,707,1145,1070,1164,731,609,7...            31985   
10035  [513,618,497,493,552,532,616,685,407,712,900,5...            22648   
...                                                  ...              ...   
19948  [305,280,267,197,490,481,484,463,401,229,157,3...            10067   
20862  [266,212,246,173,261,247,252,182,190,165,132,2...             6727   
17718  [58,41,28,47,62,50,74,48,23,34,33,40,53,53,69,...             1439   
3191   [238,239,234,158,280,361,436,343,251,194,130,2...             8328   
17524  [51,47,89,20,40,79,95,75,19,5,36,40,58,71,12,7...             1653   

       RAW_DEVICE_COUNTS  MEDIAN_DWELL  \
17609               2601     33.350000   
3175                1236    157.716667   
17751               1755     40.016667   
13849               8819     10.500000   
10035               4942    101.683333   
...                  ...           ...   
19948               2279     13.533333   
20862               1563     11.566667   
17718                416     42.866667   
3191                2040     10.966667   
17524                456      1.500000   

                                      STOPS_BY_EACH_HOUR  
17609  [0,0,23,0,0,0,0,0,0,0,0,43,23,14,19,0,15,52,22...  
3175   [33,0,0,0,0,0,0,0,0,0,22,21,0,0,21,0,20,43,20,...  
17751  [14,24,0,0,0,0,0,0,0,16,0,19,16,0,18,0,22,37,0...  
13849  [92,20,0,0,0,14,0,153,95,42,61,42,21,52,60,38,...  
10035  [0,0,0,0,0,0,0,24,17,0,58,39,40,0,62,18,40,24,...  
...                                                  ...  
19948  [2,0,0,7,4,4,11,10,16,26,15,16,44,20,17,18,22,...  
20862  [0,0,0,0,0,0,9,16,22,25,13,12,23,18,22,12,21,2...  
17718  [-2,0,0,0,0,0,6,6,3,5,5,-3,2,-3,5,10,6,10,4,-3...  
3191   [-2,0,0,0,6,5,-2,26,11,26,16,10,20,14,10,21,18...  
17524  [10,-2,-4,7,7,0,-2,0,0,0,0,-4,0,0,0,3,8,-2,0,3...  

[220684 rows x 7 columns]

Explore how to use sochastic simulation

In [39]:
print("RAW_STOP_COUNTS (billion):", month_df['RAW_STOP_COUNTS'].sum() / 1000000000)

RAW_STOP_COUNTS (billion): 4.145658773


In [40]:
df = month_df.sample(1)
df

AREA                                  DEVICE_HOME_AREAS  \
11918  170310701031  {"170310701031":13,"170310714003":13,"17031070...   

                                            STOPS_BY_DAY  RAW_STOP_COUNTS  \
11918  [115,119,103,92,123,153,136,74,211,166,93,116,...             3604   

       RAW_DEVICE_COUNTS  MEDIAN_DWELL  \
11918               2677          10.0   

                                      STOPS_BY_EACH_HOUR  
11918  [0,0,0,0,0,0,0,0,17,0,0,0,0,15,42,34,0,0,0,16,...

In [41]:
# df

In [42]:
stop_by_day = json.loads(df.iloc[0]['STOPS_BY_DAY'])
device_home_areas = json.loads(df.iloc[0]['DEVICE_HOME_AREAS'])
stops_by_each_hour = json.loads(df.iloc[0]['STOPS_BY_EACH_HOUR'])

sum(stop_by_day)
# stops_by_each_hour

3611

In [43]:
# len(device_home_areas.keys())
home_CBGs = device_home_areas.keys()
home_CBGs_device_counts = list(device_home_areas.values())
home_CBGs_device_counts_sum = sum(home_CBGs_device_counts)
print("Sum of home_area_device_counts:", home_CBGs_device_counts_sum)
raw_device = df.iloc[0]['RAW_DEVICE_COUNTS']
raw_stop_counts = df.iloc[0]['RAW_STOP_COUNTS']
stop_per_device = raw_stop_counts / raw_device
print("stop_per_device:", stop_per_device)
print("home CBG count:", len(home_CBGs))
 


Sum of home_area_device_counts: 4396
stop_per_device: 1.346283152782966
home CBG count: 1054


In [44]:
stop_home_cnt_list = []
for hour_stop in stops_by_each_hour[7:]:
    print("hour_stop:", hour_stop)
    for home_CBG, device_cnt in device_home_areas.items():
        stop_home_cnt = np.random.binomial(hour_stop, device_cnt/home_CBGs_device_counts_sum, 30)
        stop_home_cnt_list.append(stop_home_cnt)

         
    break

np.array(stop_home_cnt_list).shape
stop_home_cnt_arr = np.array(stop_home_cnt_list)
print("result shape:", stop_home_cnt_arr.shape)  # shape=30
print("simulation mean:", np.array(stop_home_cnt_list).sum(axis=0).mean())
stop_home_cnt_list[0]

hour_stop: 0
result shape: (1054, 30)
simulation mean: 0.0


array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0])

In [45]:
stop_home_cnt_list[0], stop_home_cnt_list[-1], home_CBGs_device_counts[0], home_CBGs_device_counts[-1]

(array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0]),
 array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0]),
 13,
 4)

In [46]:
stop_home_cnt_arr[:, -1].sum()
stop_home_cnt_arr.sum(axis=0).mean()


0.0

In [47]:
# home_CBGs_device_counts
# home_CBGs
# device_home_areas

In [48]:
# print(df.iloc[0]['DEVICE_HOME_AREAS'])

## 

## Check the total of the hourly stops

Usually are 98.9% of the sum of STOPS_BY_EACH_HOUR column

In [49]:
def _get_total_hourly_stop(row):
    # print(row)
    hourly_stop = json.loads(row['STOPS_BY_EACH_HOUR'])
    total_stop = sum(hourly_stop)
    return total_stop

month_df['total_hourly_stop'] = month_df.iloc[:].apply(_get_total_hourly_stop, axis=1)
ratio = month_df['total_hourly_stop'].sum() / month_df['RAW_STOP_COUNTS'].sum()

print("Ratio: total_hourly_stop / RAW_STOP_COUNTS = ", ratio)

# the total_hourly_stop is smaller than the RAW_STOP_COUNTS (98.9%)

Ratio: total_hourly_stop / RAW_STOP_COUNTS =  0.9886882101089896


## Adjust the RAW_STOP_COUNTS by dwellling_time

If a stop dwelling > 60 minutes, we add a stop to the next hours. 

May increase 25% stops.

In [50]:
print("adjusting stops by dwelling time...")
non_adjusted_dwell_time_stop_CBG_arr = ad_op.adjust_stop_by_dwelling_time(np_df=month_df.iloc[:], adjust_dwell_time=False, clean_negative=True)
print("sum of non_adjusted_dwell_time_stop_CBG_arr (billion):", non_adjusted_dwell_time_stop_CBG_arr.sum().sum()/1000000000)
print("shape of non_adjusted_dwell_time_stop_CBG_arr:", non_adjusted_dwell_time_stop_CBG_arr.shape)

adjusted_dwell_time_stop_CBG_arr = ad_op.adjust_stop_by_dwelling_time(np_df=month_df.iloc[:], adjust_dwell_time=True, clean_negative=True)

adjusted_dwell_time_stop_national_arr = adjusted_dwell_time_stop_CBG_arr.sum(axis=0)

month_df['adjusted_raw_stop'] = adjusted_dwell_time_stop_CBG_arr.sum(axis=1)
month_df[['adjusted_raw_stop', 'RAW_STOP_COUNTS']].sum() / 1000000000

print("sum of adjusted_dwell_time_stop_CBG_arr (billion):", adjusted_dwell_time_stop_CBG_arr.sum().sum()/1000000000)
print("sum of adjusted_dwell_time_stop_national_arr (billion):", adjusted_dwell_time_stop_national_arr.sum().sum()/1000000000)
print("shape of adjusted_dwell_time_stop_CBG_arr:", adjusted_dwell_time_stop_CBG_arr.shape)

adjusting stops by dwelling time...
sum of non_adjusted_dwell_time_stop_CBG_arr (billion): 4.099559846
shape of non_adjusted_dwell_time_stop_CBG_arr: (220684, 720)
sum of adjusted_dwell_time_stop_CBG_arr (billion): 5.196196864
sum of adjusted_dwell_time_stop_national_arr (billion): 5.196196864
shape of adjusted_dwell_time_stop_CBG_arr: (220684, 720)


In [51]:
# month_df

###　Save adjusted  RAW_STOP_COUNTS 

In [52]:
# CBG level
hourly_columns = [f'hour_{h+1}' for h in range(adjusted_dwell_time_stop_CBG_arr.shape[1])]
adjusted_dwelling_time_stop_CBG_df = pd.DataFrame(adjusted_dwell_time_stop_CBG_arr, columns=hourly_columns)
adjusted_dwelling_time_stop_CBG_df['CBG'] = month_df['AREA'].astype(str).str.zfill(12).to_list()
adjusted_dwelling_time_stop_CBG_df = adjusted_dwelling_time_stop_CBG_df[['CBG'] + hourly_columns]  # reorder columns
adjusted_dwelling_time_stop_CBG_df = adjusted_dwelling_time_stop_CBG_df.sort_values('CBG')

# national level
adjusted_dwelling_time_stop_natinol_df = pd.DataFrame(adjusted_dwell_time_stop_national_arr, columns=['stop'])

adjusted_dwelling_time_stop_CBG_df

CBG  hour_1  hour_2  hour_3  hour_4  hour_5  hour_6  hour_7  \
0       010010201001       0       0      23       0       0       0       0   
1       010010201002      33      33      33       0       0       0       0   
2       010010202001      14      24       0       0       0       0       0   
3       010010202002      92      20       0       0       0      14       0   
4       010010203001      33       0       0       0       0       0       0   
...              ...     ...     ...     ...     ...     ...     ...     ...   
220679  780309611001       2       0       0       7       4       4      11   
220680  780309611002       0       0       0       0       0       0       9   
220681  780309612001       2       0       0       0       0       0       6   
220682  780309612002       2       0       0       0       6       5       2   
220683  780309900000      10       2       4       7       7       0       2   

        hour_8  hour_9  hour_10  hour_11  hour_12  hour_13  hour_14  hour_15  \
0            0       0        0        0       43       23       14       19   
1            0       0        0       22       43       43       21       21   
2            0       0       16        0       19       16        0       18   
3          153      95       42       61       42       21       52       60   
4           24      41       17       58       97       79       40       62   
...        ...     ...      ...      ...      ...      ...      ...      ...   
220679      10      16       26       15       16       44       20       17   
220680      16      22       25       13       12       23       18       22   
220681       6       3        5        5        3        2        3        5   
220682      26      11       26       16       10       20       14       10   
220683       0       0        0        0        4        0        0        0   

        hour_16  hour_17  hour_18  hour_19  hour_20  hour_21  hour_22  \
0             0       15       52       22       14       20       19   
1            21       41       63       83       84       65       65   
2             0       22       37        0        0       23        0   
3            38       72       75      132       33        0       14   
4            80       58       64       76       52       60       80   
...         ...      ...      ...      ...      ...      ...      ...   
220679       18       22       33       23        2        0        1   
220680       12       21       21        5        4       12        8   
220681       10        6       10        4        3        0        5   
220682       21       18       16       18       16        0        1   
220683        3        8        2        0        3        0        0   

        hour_23  hour_24  hour_25  hour_26  hour_27  hour_28  hour_29  \
0            22        0        0       19        0       16        0   
1            63       39       19        0        0        0        0   
2            22       24        0        0        0        0        0   
3            42       22       22        0        0        0        0   
4            81       85       48       24        0        0        0   
...         ...      ...      ...      ...      ...      ...      ...   
220679        8        0        0        0        0        0        0   
220680        4        2        0        5        0        0        0   
220681        0        0        0        0        0        0        0   
220682        0       10        9        5        0        1        5   
220683        0        0        0        0        0        0        0   

        hour_30  hour_31  hour_32  hour_33  hour_34  hour_35  hour_36  \
0             0       24       23        0        0       18        0   
1             0        0       16       38      114      137      115   
2             0        0       23        0        0        0        0   
3             0        0       78       73       

In [53]:
adjusted_dwelling_time_stop_natinol_df['stop'].sum() / 10**9

5.196196864

In [54]:
# Save CSV files
adjusted_dwelling_time_stop_natinol_df.to_csv(adjusted_dwelling_time_stop_national_fname, index=False)
print("Saved national hourly stop at:", adjusted_dwelling_time_stop_national_fname)

print("Saving CBG hourly stop CSV file...")
adjusted_dwelling_time_stop_CBG_df.to_csv(adjusted_dwelling_time_stop_CBG_fname, index=False)
print("Saved CBG hourly stop at:", adjusted_dwelling_time_stop_CBG_fname)

Saved national hourly stop at: E:\OneDrive_PSU\OneDrive - The Pennsylvania State University\Research_doc\Wild_fire\hourly_map_test_2024_home_panel\adjusted_dwelling_time_stop_national_2023_06.csv
Saving CBG hourly stop CSV file...
Saved CBG hourly stop at: E:\OneDrive_PSU\OneDrive - The Pennsylvania State University\Research_doc\Wild_fire\hourly_map_test_2024_home_panel\adjusted_dwelling_time_stop_CBG_2023_06.csv


In [55]:
print("Total stops (billion):")
adjusted_dwelling_time_stop_natinol_df['stop'].sum() /  10**9, adjusted_dwelling_time_stop_CBG_df.iloc[:, 1:].sum().sum() /  10**9


Total stops (billion):


(5.196196864, 5.196196864)

In [56]:
adjusted_dwelling_time_stop_CBG_df

CBG  hour_1  hour_2  hour_3  hour_4  hour_5  hour_6  hour_7  \
0       010010201001       0       0      23       0       0       0       0   
1       010010201002      33      33      33       0       0       0       0   
2       010010202001      14      24       0       0       0       0       0   
3       010010202002      92      20       0       0       0      14       0   
4       010010203001      33       0       0       0       0       0       0   
...              ...     ...     ...     ...     ...     ...     ...     ...   
220679  780309611001       2       0       0       7       4       4      11   
220680  780309611002       0       0       0       0       0       0       9   
220681  780309612001       2       0       0       0       0       0       6   
220682  780309612002       2       0       0       0       6       5       2   
220683  780309900000      10       2       4       7       7       0       2   

        hour_8  hour_9  hour_10  hour_11  hour_12  hour_13  hour_14  hour_15  \
0            0       0        0        0       43       23       14       19   
1            0       0        0       22       43       43       21       21   
2            0       0       16        0       19       16        0       18   
3          153      95       42       61       42       21       52       60   
4           24      41       17       58       97       79       40       62   
...        ...     ...      ...      ...      ...      ...      ...      ...   
220679      10      16       26       15       16       44       20       17   
220680      16      22       25       13       12       23       18       22   
220681       6       3        5        5        3        2        3        5   
220682      26      11       26       16       10       20       14       10   
220683       0       0        0        0        4        0        0        0   

        hour_16  hour_17  hour_18  hour_19  hour_20  hour_21  hour_22  \
0             0       15       52       22       14       20       19   
1            21       41       63       83       84       65       65   
2             0       22       37        0        0       23        0   
3            38       72       75      132       33        0       14   
4            80       58       64       76       52       60       80   
...         ...      ...      ...      ...      ...      ...      ...   
220679       18       22       33       23        2        0        1   
220680       12       21       21        5        4       12        8   
220681       10        6       10        4        3        0        5   
220682       21       18       16       18       16        0        1   
220683        3        8        2        0        3        0        0   

        hour_23  hour_24  hour_25  hour_26  hour_27  hour_28  hour_29  \
0            22        0        0       19        0       16        0   
1            63       39       19        0        0        0        0   
2            22       24        0        0        0        0        0   
3            42       22       22        0        0        0        0   
4            81       85       48       24        0        0        0   
...         ...      ...      ...      ...      ...      ...      ...   
220679        8        0        0        0        0        0        0   
220680        4        2        0        5        0        0        0   
220681        0        0        0        0        0        0        0   
220682        0       10        9        5        0        1        5   
220683        0        0        0        0        0        0        0   

        hour_30  hour_31  hour_32  hour_33  hour_34  hour_35  hour_36  \
0             0       24       23        0        0       18        0   
1             0        0       16       38      114      137      115   
2             0        0       23        0        0        0        0   
3             0        0       78       73       

In [57]:
## DEVICE_HOME_AREAS column report less CBGs in RAW_DEVICE_COUNTS column :  0.8428305449098222
            ## about 85%.

RAW_DEVICE_COUNTS is large than the sum(device) in "DEVICE_HOME_AREAS", result_df['device'].sum() / result_df['RAW_DEVICE_COUNTS'].sum() = 85%

The DEVICE_HOME_AREAS column report less CBGs # thant the RAW_STOP_COUNTS column.

In [58]:
## DEVICE_HOME_AREAS column report less CBGs in RAW_DEVICE_COUNTS column :  0.8428305449098222
## about 85%.
    
def _sum_home_device_count(row):
    device_home_areas_dict = json.loads(row['DEVICE_HOME_AREAS'])
    total = sum(device_home_areas_dict.values())
    return total

month_df['sum_home_device'] = month_df.progress_apply(_sum_home_device_count, axis=1)

100%|████████████████████████████████████████████████████████████████████████| 220684/220684 [00:24<00:00, 8895.57it/s]


In [59]:
month_df

AREA                                  DEVICE_HOME_AREAS  \
17609  010010201001  {"010010201001":84,"010010208021":63,"01051031...   
3175   010010201002  {"010010201002":132,"010010209002":21,"0100102...   
17751  010010202001  {"010010208012":71,"010010202001":71,"01001020...   
13849  010010202002  {"010010205003":193,"010010205002":193,"010010...   
10035  010010203001  {"010010203001":528,"010010208024":111,"010010...   
...             ...                                                ...   
19948  780309611001  {"780309611001":39,"780309608001":21,"78030961...   
20862  780309611002  {"780309611002":48,"780309611001":44,"78030960...   
17718  780309612001  {"780309611001":21,"780309608001":20,"78030960...   
3191   780309612002  {"780309611001":28,"780309611002":23,"78030960...   
17524  780309900000  {"780309608001":11,"780309601003":6,"780309900...   

                                            STOPS_BY_DAY  RAW_STOP_COUNTS  \
17609  [286,254,196,74,246,284,287,227,272,212,249,20...             7936   
3175   [248,206,132,132,191,155,156,98,94,172,118,130...             5227   
17751  [229,61,189,267,93,72,191,112,215,177,52,152,5...             5516   
13849  [1097,900,666,861,707,1145,1070,1164,731,609,7...            31985   
10035  [513,618,497,493,552,532,616,685,407,712,900,5...            22648   
...                                                  ...              ...   
19948  [305,280,267,197,490,481,484,463,401,229,157,3...            10067   
20862  [266,212,246,173,261,247,252,182,190,165,132,2...             6727   
17718  [58,41,28,47,62,50,74,48,23,34,33,40,53,53,69,...             1439   
3191   [238,239,234,158,280,361,436,343,251,194,130,2...             8328   
17524  [51,47,89,20,40,79,95,75,19,5,36,40,58,71,12,7...             1653   

       RAW_DEVICE_COUNTS  MEDIAN_DWELL  \
17609               2601     33.350000   
3175                1236    157.716667   
17751               1755     40.016667   
13849               8819     10.500000   
10035               4942    101.683333   
...                  ...           ...   
19948               2279     13.533333   
20862               1563     11.566667   
17718                416     42.866667   
3191                2040     10.966667   
17524                456      1.500000   

                                      STOPS_BY_EACH_HOUR  total_hourly_stop  \
17609  [0,0,23,0,0,0,0,0,0,0,0,43,23,14,19,0,15,52,22...               7905   
3175   [33,0,0,0,0,0,0,0,0,0,22,21,0,0,21,0,20,43,20,...               5215   
17751  [14,24,0,0,0,0,0,0,0,16,0,19,16,0,18,0,22,37,0...               5500   
13849  [92,20,0,0,0,14,0,153,95,42,61,42,21,52,60,38,...              31794   
10035  [0,0,0,0,0,0,0,24,17,0,58,39,40,0,62,18,40,24,...              22464   
...                                                  ...                ...   
19948  [2,0,0,7,4,4,11,10,16,26,15,16,44,20,17,18,22,...               9694   
20862  [0,0,0,0,0,0,9,16,22,25,13,12,23,18,22,12,21,2...               6336   
17718  [-2,0,0,0,0,0,6,6,3,5,5,-3,2,-3,5,10,6,10,4,-3...               1115   
3191   [-2,0,0,0,6,5,-2,26,11,26,16,10,20,14,10,21,18...               8096   
17524  [10,-2,-4,7,7,0,-2,0,0,0,0,-4,0,0,0,3,8,-2,0,3...               1454   

       adjusted_raw_stop  sum_home_device  
17609               7905             2232  
3175               15645             1038  
17751               5500             1523  
13849              31794             7640  
10035              44928             4561  
...                  ...              ...  
19948               9850              870  
20862               6530              596  
17718               1573              190  
3191                8300              501  
17524               1642              238  

[220684 rows x 10 columns]

In [60]:
print("sum of home device count / raw device count:")
month_df['sum_home_device'].sum() / month_df['RAW_DEVICE_COUNTS'].sum()

sum of home device count / raw device count:


0.8377366794093563

In [61]:
(month_df['sum_home_device'] / month_df['RAW_DEVICE_COUNTS']).replace([np.inf, -np.inf], np.nan, inplace=False).dropna().mean()

0.8227918193438897

In [62]:
(month_df['RAW_DEVICE_COUNTS'] == 0).sum(), (month_df['RAW_STOP_COUNTS'] == 0).sum()

(12, 8)

### verify the relationship between raw_device_count and the sum of 720 hours.

In [63]:
print("Make a test...")


df = month_df.query("61 < MEDIAN_DWELL < 63")
# df = month_df.sample(300)

device_home_areas_test_df = ad_op.split_device_home_areas_stops(np_df=df.iloc[:])
result_df = device_home_areas_test_df.groupby('destination', as_index=False)[['stop', 'device']].sum().merge(df, left_on='destination', right_on='AREA').drop(columns='AREA')
# result_df = result_df[['destination', 'stop', 'RAW_STOP_COUNTS']]    # [['AREA', 'RAW_STOP_COUNTS', 'RAW_DEVICE_COUNTS']]


stop_diff = result_df['stop'].sum() / result_df['adjusted_raw_stop'].sum()
print("stop_diff ratio: ", stop_diff)

adjusted_diff = result_df['adjusted_raw_stop'].sum() / result_df['RAW_STOP_COUNTS'].sum()
print("adjusted stop diff ratio: ", adjusted_diff)

device_diff = result_df['device'].sum() / result_df['RAW_DEVICE_COUNTS'].sum()
print("DEVICE_HOME_AREAS column report less CBGs in RAW_DEVICE_COUNTS column （ratio）: ", device_diff)
result_df.head()

Make a test...


2138it [00:00, 3315.79it/s]


Merging columns...
stop_diff ratio:  1.0
adjusted stop diff ratio:  1.968305894356929
DEVICE_HOME_AREAS column report less CBGs in RAW_DEVICE_COUNTS column （ratio）:  1.0


destination     stop  device  \
0  010090501013  38650.0  2520.0   
1  010150012022  36258.0  4009.0   
2  010150014001   8668.0  1038.0   
3  010179542001  45608.0  1915.0   
4  010310105003  12168.0  1524.0   

                                   DEVICE_HOME_AREAS  \
0  {"010090501013":417,"010090501014":111,"010090...   
1  {"010150012022":394,"010150011001":125,"010150...   
2  {"010150014001":45,"011210102011":22,"01015000...   
3  {"010179542001":461,"010179544001":61,"0108104...   
4  {"010310105003":15,"010310105002":4,"010399623...   

                                        STOPS_BY_DAY  RAW_STOP_COUNTS  \
0  [735,673,282,716,643,436,528,506,622,810,906,3...            19604   
1  [626,761,514,570,757,820,1163,850,685,495,487,...            18393   
2  [128,125,161,185,79,162,92,116,203,92,107,113,...             4472   
3  [559,495,643,1371,984,647,725,778,797,374,627,...            23206   
4  [188,184,109,149,149,168,91,150,148,361,258,13...             6188   

   RAW_DEVICE_COUNTS  MEDIAN_DWELL  \
0               2520     62.466667   
1               4009     61.833333   
2               1038     61.933333   
3               1915     62.316667   
4               1524     61.266667   

                                  STOPS_BY_EACH_HOUR  total_hourly_stop  \
0  [14,22,0,15,0,0,0,0,0,0,32,56,0,74,19,80,185,5...              19325   
1  [0,20,0,56,0,0,37,32,13,38,14,16,13,16,54,32,3...              18129   
2  [18,0,0,0,0,0,0,0,0,0,0,0,35,0,0,0,0,0,0,15,0,...               4334   
3  [14,0,0,0,21,21,0,36,68,12,33,0,0,19,36,13,34,...              22804   
4  [0,0,0,0,0,0,12,0,0,0,19,8,0,0,8,29,29,41,0,0,...               6084   

   adjusted_raw_stop  sum_home_device  
0              38650             2366  
1              36258             3542  
2               8668              882  
3              45608             2018  
4              12168              574

In [64]:
# device_home_areas_df

### Save the a single CSV for further use, only keep the non-dictionary columns.

In [65]:
non_dict_columns = ['AREA', 
                    "STOPS_BY_DAY", 
                    "RAW_STOP_COUNTS", 
                    "RAW_DEVICE_COUNTS", 
                    "MEDIAN_DWELL", 
                    "STOPS_BY_EACH_HOUR",
                   'adjusted_raw_stop',
                   ]
# save CSV file
# month_df[non_dict_columns].to_csv(no_dict_colum_fname, index=False)
print("Combined CSV wihouth dictionary columns was saved at:", no_dict_colum_fname)

Combined CSV wihouth dictionary columns was saved at: E:\OneDrive_PSU\OneDrive - The Pennsylvania State University\Research_doc\Wild_fire\hourly_map_test_2024_home_panel\NP_dict_column_2023_06.csv


## Split the DEVICE_HOME_AREAS column

Some CBGs are not reported because their visitors < 4, we add them according to the RAW_DEVICE_COUNTS

Note: need > 30 GB memory

In [ ]:
## DEVICE_HOME_AREAS column report less CBGs in RAW_DEVICE_COUNTS column :  0.8428305449098222
## about 85%.
# assumption: in a destination CBG, each device contribute the same stop counts in a month.

device_home_areas_df = ad_op.split_device_home_areas_stops(np_df=month_df.iloc[:])

# print("Sorting...")
## device_home_areas_df = device_home_areas_df.sort_values(['origin', 'destination'])
## device_home_areas_df = device_home_areas_df.sort_values(['origin', 'device'])
# device_home_areas_df['origin'] = device_home_areas_df['origin'].astype(str).str.zfill(12)
# device_home_areas_df['destination'] = device_home_areas_df['destination'].astype(str).str.zfill(12)
## device_home_areas_df['device'] = device_home_areas_df['device'].astype(int)

print(f"Saving at: {device_home_areas_fname}...")
device_home_areas_df.to_csv(device_home_areas_fname, index=False)
print("Done")
device_home_areas_df

14941it [00:06, 2693.37it/s]

In [ ]:
print("summary (billion):")
device_home_areas_df[['device', 'stop']].sum() / 10**9, month_df[['RAW_DEVICE_COUNTS', 'RAW_STOP_COUNTS', 'adjusted_raw_stop']].sum() / 10**9

### Save aggregation for origin and destination distribution

In [ ]:
# Save aggregation for origin and destination distribution
CBG_device_stop_destination_distribution_df = device_home_areas_df.groupby('destination')[['device', 'stop']].sum()
CBG_device_stop_destination_distribution_df.to_csv(CBG_device_stop_destination_distribution_fname)

CBG_device_stop_origin_distribution_df = device_home_areas_df.groupby('origin')[['device', 'stop']].sum()
CBG_device_stop_origin_distribution_df.to_csv(CBG_device_stop_origin_distribution_fname)

CBG_device_stop_origin_distribution_df

In [ ]:
CBG_device_stop_destination_distribution_df#.reset_index()

In [ ]:
print("Summary (destination and origin suppose to be the same):")
CBG_device_stop_destination_distribution_df.sum()  / 10**9, CBG_device_stop_origin_distribution_df.sum()  / 10**9

## Generate person_per_device (i.e., device sampling rate) file for origin CBGs

In [ ]:
ACS_df = pd.read_csv(ACS_file)
ACS_df["CBG_FIPS"] = ACS_df["fips"].astype(str).str.zfill(12)
print("Total US population (million):",  ACS_df['totalpopulation'].sum() / 10**6)

In [ ]:
print("Loading home panel file:", home_panel_fname)
panel_df = pd.read_csv(home_panel_fname)
panel_df

In [ ]:
monthly_device_cnt_df = panel_df.query(f"ISO_COUNTRY_CODE == 'US' ").groupby(['YEAR', 'MON'], as_index=True)['NUMBER_DEVICES_RESIDING'].sum() / 10**6
print("monthly_device_cnt_df   (million):" )
monthly_device_cnt_df.reset_index()

The NUMBER_DEVICES_RESIDING surged from 45 million to 900 million, and the sampling rate cannot be obatain, because it exceeds the US population.

In [ ]:
target_panel_df = panel_df.query(f"YEAR == {year} and MON == {month} and ISO_COUNTRY_CODE == 'US' ").sort_values('CENSUS_BLOCK_GROUP')  # 
# target_panel_df = panel_df.query(f"YEAR == {year} and MON == {6} and ISO_COUNTRY_CODE == 'US' ").sort_values('CENSUS_BLOCK_GROUP')  # 
print("Total US NUMBER_DEVICES_RESIDING (million):",  target_panel_df['NUMBER_DEVICES_RESIDING'].sum() / 10**6)
target_panel_df

In [ ]:
panel_ACS_df = target_panel_df.merge(ACS_df[['CBG_FIPS', 'totalpopulation']], left_on='CENSUS_BLOCK_GROUP', right_on='CBG_FIPS')
panel_ACS_df = panel_ACS_df.dropna(subset='NUMBER_DEVICES_RESIDING')
panel_ACS_df['NUMBER_DEVICES_RESIDING'] = panel_ACS_df['NUMBER_DEVICES_RESIDING'].astype(int)
panel_ACS_df = panel_ACS_df.sort_values('CBG_FIPS')
panel_ACS_df['person_per_device'] = panel_ACS_df['totalpopulation'] / panel_ACS_df['NUMBER_DEVICES_RESIDING']
panel_ACS_df['sampling_rate'] = panel_ACS_df['NUMBER_DEVICES_RESIDING'] / panel_ACS_df['totalpopulation']

print("Data cover US population (million) :", panel_ACS_df['totalpopulation'].sum() / 10**6)
panel_ACS_df

In [ ]:
fig, ax = plt.subplots()
# month = 6
ax.set_title(f"Device sampling rate histogram ({year}-{month:02})")
ax.set_xlabel("Sampling rate")
ax.set_ylabel("CBG count")

panel_ACS_df.query("totalpopulation > 0  ")['sampling_rate'].dropna().hist(ax=ax, bins=np.array(range(0, 200,1))/100)

In [ ]:
# get people_per_device of a county

panel_ACS_df['county_FIPS'] = panel_ACS_df['CBG_FIPS'].str[:5]
sampling_rate_county_df = panel_ACS_df.groupby(['county_FIPS'], as_index=False).agg(    # , 'YEAR', "MON", "REGION", "ISO_COUNTRY_CODE"
    county_device=('NUMBER_DEVICES_RESIDING', 'sum'),
    county_device_daytime=('NUMBER_DEVICES_RESIDING', 'sum'),
    county_device_population=('totalpopulation', 'sum'),
    ) 

sampling_rate_county_df['person_per_device_county'] = sampling_rate_county_df['county_device_population'] /  sampling_rate_county_df['county_device']

print("National device and population count:\n")
print(sampling_rate_county_df[['county_device', 'county_device_population']].sum())

people_per_device_df = panel_ACS_df.merge(sampling_rate_county_df, left_on='county_FIPS', right_on='county_FIPS')
people_per_device_df.to_csv(person_per_device_fname, index=False)
people_per_device_df

## Compute the inbounds for each destination CBG

merge the device_home_area and person_per_device, then groupby the destination

In [ ]:
if 'device_home_areas_df' not in locals():
    device_home_areas_df = pd.read_csv(device_home_areas_fname, dtype={"origin":str, "destination":str})
device_home_areas_df

In [ ]:
print("summary of device, stop (billion):")
device_home_areas_df[['device', 'stop']].sum() / 10**9

# for 2023-06
# summary of device, stop (billion):
# device    0.887546
# stop      5.196190

In [ ]:
# people_per_device_df = pd.read_csv(person_per_device_fname, index=False)

print("Loading people_per_device_df from:", person_per_device_fname)
person_per_device_df = pd.read_csv(person_per_device_fname, dtype={"CBG_FIPS": str})

print("Total totalpopulation (billion):", person_per_device_df['totalpopulation'].sum() / 10**9)
person_per_device_df

In [ ]:
origin_person_per_device_df = device_home_areas_df.merge(person_per_device_df[['CBG_FIPS', 'person_per_device', 'person_per_device_county']], left_on="origin", right_on="CBG_FIPS")
origin_person_per_device_df

In [ ]:
print("Total outbound (actual visitors, billion):")
(origin_person_per_device_df['stop'] * origin_person_per_device_df['person_per_device']).sum() / 10**9

In [ ]:
# IMPORTANT! !!!!!!!!!!!
origin_person_per_device_df['person_CBG_rate'] = origin_person_per_device_df['device'] * origin_person_per_device_df['person_per_device']  
# ’person_CBG_rate‘ means using the CBG device sampling rate
origin_person_per_device_df['person_county_rate'] = origin_person_per_device_df['device'] * origin_person_per_device_df['person_per_device_county']
# ’person_per_device_county‘ means using the county device sampling rate

# origin_person_per_device_df[['person_CBG_rate','person_county_rate']].sum() / 10**9

In [ ]:
# use the device count and person_per_device of the origin CBG to compute the total stops of the month.
# actually, the total visitor count (leave their homes) at every hour in a month
print("Total visitors per hour in this month:")
(origin_person_per_device_df['stop'] * origin_person_per_device_df['person_per_device']).sum() / 10**9

In [ ]:
# IMPORTANT! !!!!!!!!!!!
# we assume the stopped devices in destination CBG is a combination of those from home CBGs, which means that the people represent by a destination CBG devie is also a combination of CBG device.
# To get this combinated amount:
# 1. aggreate the people from each home CBG. (assume that each device represen the same amount of people all the time, a device contribute the same amount of stops alll the time!)

def get_inbound_person_per_device(df):
    inboud_per_per_device_CBG_rate = df['person_CBG_rate'].sum() /  df['device'].sum()
    inboud_per_per_device_county_rate = df['person_county_rate'].sum() /  df['device'].sum()
    return {"inboud_per_device_CBG_rate": inboud_per_per_device_CBG_rate, "inboud_per_device_county_rate": inboud_per_per_device_county_rate}

CBG_destination_inbound_per_device = origin_person_per_device_df.iloc[:].groupby('destination').progress_apply(get_inbound_person_per_device)
CBG_destination_inbound_per_device_df = pd.DataFrame(CBG_destination_inbound_per_device.tolist(), index=CBG_destination_inbound_per_device.index)
CBG_destination_inbound_per_device_df.to_csv(CBG_destination_inbound_per_device_fname)
print(CBG_destination_inbound_per_device_df.mean())
CBG_destination_inbound_per_device_df

In [ ]:
print("Loading CBG_destination_inbound_per_device_df from:", CBG_destination_inbound_per_device_fname)
CBG_destination_inbound_per_device_df = pd.read_csv(CBG_destination_inbound_per_device_fname, dtype={"destination": str}).set_index('destination')

print("CBG_destination_inbound_per_device_df.mean(): \n", CBG_destination_inbound_per_device_df.mean() )
CBG_destination_inbound_per_device_df

In [ ]:
print("Loading adjusted_dwelling_time_stop_CBG_df from:", adjusted_dwelling_time_stop_CBG_fname)
adjusted_dwelling_time_stop_CBG_df = pd.read_csv(adjusted_dwelling_time_stop_CBG_fname, dtype={"CBG": str}).set_index('CBG')

print("Total adjusted hourly stops (billion):", adjusted_dwelling_time_stop_CBG_df.iloc[:, 1:].sum().sum() / 10**9)
adjusted_dwelling_time_stop_CBG_df

In [ ]:
adjusted_dwelling_time_stop_inbound_per_device_CBG_df = adjusted_dwelling_time_stop_CBG_df.merge(CBG_destination_inbound_per_device_df,left_index=True, right_index=True)
adjusted_dwelling_time_stop_inbound_per_device_CBG_df.index.name = 'CBG'
adjusted_dwelling_time_stop_inbound_per_device_CBG_df

In [ ]:
hour_cnt = adjusted_dwelling_time_stop_CBG_df.shape[1]
print("hour_cnt:", hour_cnt)

In [ ]:
# inbound_hourly_df.iloc[:, hour]
# inbound_hourly_df.loc[:, 'inboud_per_per_device_CBG_rate']

In [ ]:

inbound_hourly_df = adjusted_dwelling_time_stop_inbound_per_device_CBG_df.astype(np.float64).copy()

for hour in tqdm(range(hour_cnt)):
    # inbound_hourly_df.iloc[:, hour] = inbound_hourly_df.iloc[:, hour] * inbound_hourly_df.loc[:, 'inboud_per_per_device_CBG_rate']   # ？ 70.4215236989238 billion
    inbound_hourly_df.iloc[:, hour] = inbound_hourly_df.iloc[:, hour] * inbound_hourly_df.loc[:, 'inboud_per_device_county_rate']   # ？ 56.34291367286323 billion
print("Total restored inbound (billion) using per_device_county_rate:",  inbound_hourly_df.iloc[:, :hour].sum().sum() / 10**9)

inbound_hourly_df = adjusted_dwelling_time_stop_inbound_per_device_CBG_df.astype(np.float64).copy()

for hour in tqdm(range(hour_cnt)):
    inbound_hourly_df.iloc[:, hour] = inbound_hourly_df.iloc[:, hour] * inbound_hourly_df.loc[:, 'inboud_per_device_CBG_rate']   # 70.4215236989238 billion
    # inbound_hourly_df.iloc[:, hour] = inbound_hourly_df.iloc[:, hour] * inbound_hourly_df.loc[:, 'inboud_per_per_device_county_rate']   # 56.34291367286323 billion
print("Total restored inbound (billion) using per_device_CBG_rate:",  inbound_hourly_df.iloc[:, :hour].sum().sum() / 10**9)

print("We will use the larger resutls: per_device_CBG_rate.")

print("Saving inbound_hourly_df at:", inbound_hourly_df_fname)
inbound_hourly_df.to_csv(inbound_hourly_df_fname)
inbound_hourly_df

# Total restored inbound (billion) using per_device_county_rate: 51.178735299160714
# Total restored inbound (billion) using per_device_CBG_rate:    61.49318556384339


In [ ]:
# Get the hourly aggregated inbounds for all CBGs
inbound_hourly_df.iloc[:, :hour_cnt].sum().sum() / 10**9

In [ ]:
hourly_agregated_inbound_df = pd.DataFrame(inbound_hourly_df.iloc[:, :hour_cnt].sum(), columns=['inboud']).reset_index()[['inboud']]
hourly_agregated_inbound_df.to_csv(hourly_agregated_inbound_fname, index=False)
print("Save hourly_agregated_inbound_df at:", hourly_agregated_inbound_fname)

CBG_agregated_inbound_df = pd.DataFrame(inbound_hourly_df.iloc[:, :hour_cnt].sum(axis=1), columns=['inboud'])
CBG_agregated_inbound_df.to_csv(CBG_agregated_inbound_fname)
print("Save CBG_agregated_inbound_df at:", CBG_agregated_inbound_fname)

In [ ]:
hourly_agregated_inbound_df

In [ ]:
CBG_agregated_inbound_df['inboud'].sum() / 10**9

In [ ]:
people_per_device_df = pd.read_csv(person_per_device_fname, dtype={'CBG_FIPS': str})
people_per_device_df

In [ ]:
# Load flow data (month)
# device_home_areas_df = pd.read_csv(device_home_areas_fname, dtype={'origin':str, 'destination':str})
device_home_areas_df

In [ ]:
device_home_areas_person_per_device_df = device_home_areas_df.merge(people_per_device_df[['CBG_FIPS', 'person_per_device', 'person_per_device_county']], left_on='origin', right_on='CBG_FIPS')
device_home_areas_person_per_device_df

In [ ]:
device_home_areas_person_per_device_df['stop'].sum() / 10**9, device_home_areas_person_per_device_df['device'].sum() / 10**9

In [ ]:
# Not correct
# device_home_areas_person_per_device_df['outbound'] = device_home_areas_person_per_device_df['stop'] * device_home_areas_person_per_device_df['person_per_device']
# device_home_areas_person_per_device_df['outbound'].sum()  / 10**9

In [ ]:
no_dict_colum_month_df = pd.read_csv(no_dict_colum_fname, dtype={"AREA":str})
no_dict_colum_month_df['destination'] = no_dict_colum_month_df['AREA'].astype(str).str.zfill(12)
no_dict_colum_month_df

In [ ]:
no_dict_colum_month_df['adjusted_raw_stop'].sum() / 10**9

In [ ]:
# device_home_areas_person_per_device_df.groupby('destination', as_index=False)#[['inbound']] .sum()
device_home_areas_person_per_device_df

In [ ]:
destination_CBG_inbound_df = device_home_areas_person_per_device_df.groupby('destination', as_index=False)[['stop']].sum()#.sum()
destination_CBG_inbound_df['destination'] = destination_CBG_inbound_df['destination'].astype(str).str.zfill(12)

destination_CBG_inbound_device_device_count_df = destination_CBG_inbound_df.merge(no_dict_colum_month_df[['destination', 'RAW_STOP_COUNTS', 'RAW_DEVICE_COUNTS', 'MEDIAN_DWELL']],
                                                                                 left_on="destination",
                                                                                 right_on="destination")

destination_CBG_inbound_device_device_count_df['inbound_per_stop'] = destination_CBG_inbound_device_device_count_df['stop'] / destination_CBG_inbound_device_device_count_df['RAW_STOP_COUNTS']
destination_CBG_inbound_device_device_count_df

In [ ]:
destination_CBG_inbound_device_device_count_df['inbound'] = destination_CBG_inbound_device_device_count_df['stop'] * destination_CBG_inbound_device_device_count_df['inbound_per_stop']


In [ ]:
destination_CBG_inbound_device_device_count_df['inbound'].sum() / 1000000000, destination_CBG_inbound_device_device_count_df['RAW_STOP_COUNTS'].sum() / 1000000000

In [ ]:
destination_CBG_inbound_device_device_count_df.to_csv(CBG_destination_inbound_fname, index=False)
print("Saved destination_CBG_inbound_device_device_count_df as:", CBG_destination_inbound_fname)
destination_CBG_inbound_device_device_count_df

In [ ]:
destination_CBG_inbound_device_device_count_df['inbound'].sum() / 1000000000

## Compute the hourly inbound (not used, not correct)

In [ ]:
# destination_CBG_inbound_df = pd.read_csv(CBG_destination_inbound_fname, dtype={'destination':str})
# destination_CBG_inbound_df

In [ ]:
# destination_CBG_inbound_df.sample(30)['inbound_per_stop']#.mean()

In [ ]:
# print("Loading in adjusted_dwelling_time_stop_CBG_df from:", adjusted_dwelling_time_stop_CBG_fname)
# adjusted_dwelling_time_stop_CBG_df = pd.read_csv(adjusted_dwelling_time_stop_CBG_fname, dtype={'CBG':str})
# print("Total adjusted observed stop (billion):", adjusted_dwelling_time_stop_CBG_df.iloc[:, 1:].sum().sum() / 1000000000)
# adjusted_dwelling_time_stop_CBG_df

In [ ]:
# hour_cnt = adjusted_dwelling_time_stop_CBG_df.shape[1] - 1
# print("hour_cnt:", hour_cnt)

In [ ]:
# adjusted_stop_inbound_per_stop_CBG_df = adjusted_dwelling_time_stop_CBG_df.merge(destination_CBG_inbound_df[['destination','inbound_per_stop']],
#                                                                                 left_on="CBG",
#                                                                                 right_on="destination",
#                                                                                 )
# adjusted_stop_inbound_per_stop_CBG_df

In [ ]:
# hourly_inbound_df = adjusted_stop_inbound_per_stop_CBG_df.copy()
# for h in range(hour_cnt):
#     hourly_inbound_df.iloc[:, h + 1] = hourly_inbound_df.iloc[:, h + 1] * hourly_inbound_df['inbound_per_stop']  

# hourly_inbound_df = hourly_inbound_df.iloc[:, :-2]
# print("Total inbound (billion):", hourly_inbound_df.iloc[:, 1:].sum().sum() / 1000000000)
# hourly_inbound_df

In [ ]:
# hourly_inbound_df.iloc[:, 1:].sum().sum()

# IPF  v2

### Restored the outbound according to device sampling rate

In [ ]:
if  'CBG_device_stop_origin_distribution_df' not in locals():
    CBG_device_stop_origin_distribution_df = pd.read_csv(CBG_device_stop_origin_distribution_fname)
print("Total observed stops (billion): ", CBG_device_stop_origin_distribution_df['stop'].sum() / 1000000000)
CBG_device_stop_origin_distribution_df

In [ ]:
people_per_device_df = pd.read_csv(person_per_device_fname, dtype={'CBG_FIPS': str})
people_per_device_df

In [ ]:
CBG_origin_stop_distribution_sampling_rate_df = CBG_device_stop_origin_distribution_df.reset_index().merge(people_per_device_df, left_on='origin', right_on='CBG_FIPS')
CBG_origin_stop_distribution_sampling_rate_df

Because "stop" is the combination of visitors from various HOME CBGs! So, we need to compute the inbound. 

The outbound compuated below can be view as the distribution only.

In [ ]:
person_per_device_national = CBG_origin_stop_distribution_sampling_rate_df['totalpopulation'].sum() / CBG_origin_stop_distribution_sampling_rate_df['NUMBER_DEVICES_RESIDING'].sum()
CBG_origin_stop_distribution_sampling_rate_df['restored_outbound_count'] = CBG_origin_stop_distribution_sampling_rate_df['stop'] * person_per_device_national
restored_total_restored_outbound_count = CBG_origin_stop_distribution_sampling_rate_df['restored_outbound_count'].sum() /  10**9
print("Restored total outbound distribution (billion) using national sampling rate:", restored_total_restored_outbound_count)  # 37.44716723845148 billion

CBG_origin_stop_distribution_sampling_rate_df['restored_outbound_count'] = CBG_origin_stop_distribution_sampling_rate_df['stop'] * CBG_origin_stop_distribution_sampling_rate_df['person_per_device_county']
restored_total_restored_outbound_count = CBG_origin_stop_distribution_sampling_rate_df['restored_outbound_count'].sum() /  10**9
print("Restored total outbound distribution (billion) using county sampling rate:", restored_total_restored_outbound_count)  # 50.96843487672891 billion

CBG_origin_stop_distribution_sampling_rate_df['restored_outbound_count'] = CBG_origin_stop_distribution_sampling_rate_df['stop'] * CBG_origin_stop_distribution_sampling_rate_df['person_per_device']
restored_total_restored_outbound_count = CBG_origin_stop_distribution_sampling_rate_df['restored_outbound_count'].sum() / 10**9
CBG_origin_stop_distribution_sampling_rate_df.to_csv(CBG_origin_outbound_distribution_sampling_rate_fname, index=False)
print("Restored total outbound distribution (billion) using CBG sampling rate:", restored_total_restored_outbound_count)  # 61.19985609714293 billion

print("Saved CBG_origin_outbound_distribution_sampling_rate at:", CBG_origin_outbound_distribution_sampling_rate_fname)

In [ ]:
# use the total_inboud to calibrate the outbound

# restored_outbound_count

## Load data

In [ ]:
hourly_agregated_inbound_df = pd.read_csv(hourly_agregated_inbound_fname)
total_inbound = hourly_agregated_inbound_df['inboud'].sum()
print("total_inbound (billion):", total_inbound / 10**9)

In [ ]:
# hourly_agregated_inbound_df

In [ ]:
CBG_restored_origin_outbound_df = pd.read_csv(CBG_origin_outbound_distribution_sampling_rate_fname)
total_outbound = CBG_restored_origin_outbound_df['restored_outbound_count'].sum()
print("Total outbound distribution (billions): ",  total_outbound / 1000000000)


factor = total_inbound / total_outbound
print("factor:", factor)


CBG_restored_origin_outbound_df['restored_outbound_count'] = CBG_restored_origin_outbound_df['restored_outbound_count'] * factor

print("Total outbound distribution (billions): ",   CBG_restored_origin_outbound_df['restored_outbound_count'].sum() / 10**9)

In [ ]:
CBG_restored_origin_outbound_df.shape
CBG_restored_origin_outbound_df


In [ ]:
# Load 

In [ ]:
# adjusted_dwelling_time_stop_natinol_df = pd.read_csv(adjusted_dwelling_time_stop_natinol_fname)
# total_adjusted_stops = adjusted_dwelling_time_stop_natinol_df['stop'].sum()
# print("total_adjusted_stops (billion):", total_adjusted_stops / 1000000000)

# # covnert distribution to acution outbound
# outbound_per_stop = total_outbound / total_adjusted_stops 

# adjusted_dwelling_time_stop_natinol_df['stop'] *= outbound_per_stop
# print("total outbouds (billion):", adjusted_dwelling_time_stop_natinol_df['stop'].sum() / 1000000000)

### start IPF

In [ ]:
# !conda install ipfn
# import ipfn

In [ ]:
import numpy as np
from ipfn import ipfn
# from ipfn import ipfn

restored_CBG_outbound_arr = CBG_restored_origin_outbound_df['restored_outbound_count'].to_numpy()
restored_hourly_outbound_arr = hourly_agregated_inbound_df['inboud'].to_numpy()  # 

aggregates = [restored_CBG_outbound_arr, restored_hourly_outbound_arr]
dimensions = [[0], [1]]

seed_arr = np.ones((len(restored_CBG_outbound_arr), len(restored_hourly_outbound_arr)))

IPF = ipfn.ipfn(seed_arr, aggregates, dimensions, convergence_rate=1e-3)
m = IPF.iteration()
print(m.shape)

In [ ]:
print("Error of IPF:")
print((m.sum(axis=1) - restored_CBG_outbound_arr).sum(), (m.sum(axis=0) - restored_hourly_outbound_arr).sum())

In [ ]:
(m.sum(axis=1) - restored_CBG_outbound_arr).max(), (m.sum(axis=0) - restored_hourly_outbound_arr).max()

In [ ]:
(m.sum(axis=1) - restored_CBG_outbound_arr).min(), (m.sum(axis=0) - restored_hourly_outbound_arr).min()

In [ ]:
# Save IPF results
hourly_columns = [f'hour_{h+1}' for h in range(restored_hourly_outbound_arr.shape[0])]
IPF_df = pd.DataFrame(m, columns=hourly_columns)
print("Total outbound visitor (billion):", IPF_df.sum().sum() / 1000000000)

 
IPF_df['CBG'] = CBG_restored_origin_outbound_df['origin'].astype(str).str.zfill(12)
IPF_df = IPF_df[['CBG'] + hourly_columns]

print("Saving IPF at:", IPF_results_fname)
IPF_df.to_csv(IPF_results_fname, index=False)
print("Saved IPF_results at:", IPF_results_fname)
IPF_df

##	Adjust stop count by dwelling time

Load the raw stop data and Neighorhood Patterns table.

In [ ]:
# Load the raw stop data and Neighorhood Patterns table.

# device_home_areas_df = pd.read_csv(device_home_areas_fname)
# print("Loading no_dict_colum CSV file:", no_dict_colum_fname)
# no_dict_colum_df = pd.read_csv(no_dict_colum_fname)

# no_dict_colum_df

In [ ]:
# non_adjusted_dwell_time_stop_CBG_arr = ad_op.adjust_stop_by_dwelling_time(np_df=no_dict_colum_df.iloc[:], adjust_dwell_time=False, clean_negative=True)
# print("sum of non_adjusted_dwell_time_stop_CBG_arr (million):", non_adjusted_dwell_time_stop_CBG_arr.sum().sum()/1000000)
# print("shape of non_adjusted_dwell_time_stop_CBG_arr:", non_adjusted_dwell_time_stop_CBG_arr.shape)

In [ ]:
# adjusted_dwell_time_stop_CBG_arr = ad_op.adjust_stop_by_dwelling_time(np_df=no_dict_colum_df.iloc[:], adjust_dwell_time=True, clean_negative=True)

# adjusted_dwell_time_stop_national_arr = adjusted_dwell_time_stop_CBG_arr.sum(axis=0)

# print("sum of adjusted_dwell_time_stop_CBG_arr (million):", adjusted_dwell_time_stop_CBG_arr.sum().sum()/1000000)
# print("sum of adjusted_dwell_time_stop_national_arr (million):", adjusted_dwell_time_stop_national_arr.sum().sum()/1000000)
# print("shape of adjusted_dwell_time_stop_CBG_arr:", adjusted_dwell_time_stop_CBG_arr.shape)

In [ ]:
# hourly_columns
# adjusted_dwell_time_stop_national_arr.shape

Convert numpy array to DataFrame

In [ ]:
# convert numpy array to DataFrame
# CBG level
# hourly_columns = [f'hour_{h+1}' for h in range(adjusted_dwell_time_stop_CBG_arr.shape[1])]
# adjusted_dwelling_time_stop_CBG_df = pd.DataFrame(adjusted_dwell_time_stop_CBG_arr, columns=hourly_columns)
# adjusted_dwelling_time_stop_CBG_df['CBG'] = no_dict_colum_df['AREA'].astype(str).str.zfill(12).to_list()
# adjusted_dwelling_time_stop_CBG_df = adjusted_dwelling_time_stop_CBG_df[['CBG'] + hourly_columns]  # reorder columns
# adjusted_dwelling_time_stop_CBG_df = adjusted_dwelling_time_stop_CBG_df.sort_values('CBG')


# # national level
# adjusted_dwelling_time_stop_natinol_df = pd.DataFrame(adjusted_dwell_time_stop_national_arr, columns=['stop'])

# adjusted_dwelling_time_stop_CBG_df

In [ ]:
# print("Adjustted total stop (billion) by dwelling time from CBG and national level (should be the same):")
# print(adjusted_dwelling_time_stop_CBG_df[hourly_columns].sum().sum() / 1000000000,  adjusted_dwelling_time_stop_natinol_df.sum().sum()/ 1000000000)

In [ ]:
# # Save CSV files
# adjusted_dwelling_time_stop_natinol_df.to_csv(adjusted_dwelling_time_stop_natinol_fname, index=False)
# print("Saved national hourly stop at:", adjusted_dwelling_time_stop_natinol_fname)

# adjusted_dwelling_time_stop_CBG_df.to_csv(adjusted_dwelling_time_stop_CBG_fname, index=False)
# print("Saved CBG hourly stop at:", adjusted_dwelling_time_stop_CBG_fname)

###  Convert observed stop counts to actual visitor counts.

#### compute the person_per_device at CBG and county level

Load ACS and home panel data, merge them into device sampling rate table at CBG and county level

In [ ]:
# ACS_df = pd.read_csv(ACS_file)
# ACS_df["CBG_FIPS"] = ACS_df["fips"].astype(str).str.zfill(12)
# print("Total US population (million):",  ACS_df['totalpopulation'].sum() / 1000000)

In [ ]:
# print("Loading home panel file:", home_panel_fname)
# panel_df = pd.read_csv(home_panel_fname)
# target_panel_df = panel_df.query(f"YEAR == {year} and MON == {int(month)} and ISO_COUNTRY_CODE == 'US' ")
# target_panel_df

In [ ]:
# panel_ACS_df = target_panel_df.merge(ACS_df[['CBG_FIPS', 'totalpopulation']], left_on='CENSUS_BLOCK_GROUP', right_on='CBG_FIPS')
# panel_ACS_df = panel_ACS_df.dropna(subset='NUMBER_DEVICES_RESIDING')
# panel_ACS_df['NUMBER_DEVICES_RESIDING'] = panel_ACS_df['NUMBER_DEVICES_RESIDING'].astype(int)
# panel_ACS_df = panel_ACS_df.sort_values('CBG_FIPS')
# panel_ACS_df['person_per_device'] = panel_ACS_df['totalpopulation'] / panel_ACS_df['NUMBER_DEVICES_RESIDING']
# print("Data cover US population:", panel_ACS_df['totalpopulation'].sum())
# panel_ACS_df

In [ ]:
# # get people_per_device of a county

# panel_ACS_df['county_FIPS'] = panel_ACS_df['CBG_FIPS'].str[:5]
# sampling_rate_county_df = panel_ACS_df.groupby(['county_FIPS'], as_index=False).agg(    # , 'YEAR', "MON", "REGION", "ISO_COUNTRY_CODE"
#     county_device=('NUMBER_DEVICES_RESIDING', 'sum'),
#     county_device_daytime=('NUMBER_DEVICES_RESIDING', 'sum'),
#     county_device_population=('totalpopulation', 'sum'),
#     ) 

# sampling_rate_county_df['person_per_device_county'] = sampling_rate_county_df['county_device_population'] /  sampling_rate_county_df['county_device']

# print("National device and population count:\n")
# print(sampling_rate_county_df[['county_device', 'county_device_population']].sum())

# people_per_device_df = panel_ACS_df.merge(sampling_rate_county_df, left_on='county_FIPS', right_on='county_FIPS')
# people_per_device_df.to_csv(person_per_device_fname, index=False)
# people_per_device_df

#### restore the observated stop count to actual visitor count

In [ ]:
# people_per_device_df = pd.read_csv(person_per_device_fname, dtype={'CBG_FIPS': str})
# adjusted_dwelling_time_stop_CBG_df = pd.read_csv(adjusted_dwelling_time_stop_CBG_fname, dtype={'CBG':str})

In [ ]:
# adjusted_dwelling_time_stop_CBG_df
# people_per_device_df

In [ ]:
# adjusted_dwelling_time_stop_CBG_df.shape

## Get the total outgoer

In [ ]:
# device_home_areas_df = pd.read_csv(device_home_areas_fname)
# device_home_areas_df

In [ ]:
# # Assume a stop is a device
# hour_cnt = adjusted_dwelling_time_stop_CBG_df.shape[1] - 1


# hourly_stop_sampling_rate_df = adjusted_dwelling_time_stop_CBG_df.merge(people_per_device_df, left_on='CBG', right_on='CBG_FIPS')
# national_people_per_device = hourly_stop_sampling_rate_df['totalpopulation'].sum() / hourly_stop_sampling_rate_df['NUMBER_DEVICES_RESIDING'].sum()

# print("hour_cnt:", hour_cnt)
# print("national_people_per_device:", national_people_per_device)

# for h in range(hour_cnt):
#     hourly_stop_sampling_rate_df.iloc[:, h + 1] = hourly_stop_sampling_rate_df.iloc[:, h + 1] * hourly_stop_sampling_rate_df['person_per_device']   # CBG leval   # BUG: person_per_device is for home(origin), not for destination!!!!!!!!!!!!
#     # hourly_stop_sampling_rate_df.iloc[:, h + 1] = hourly_stop_sampling_rate_df.iloc[:, h + 1] * national_people_per_device    # national leval
#     # hourly_stop_sampling_rate_df.iloc[:, h + 1] = hourly_stop_sampling_rate_df.iloc[:, h + 1] * hourly_stop_sampling_rate_df['person_per_device_county']   # county level
    

# ## used      CBG level sampling rate: 51028323628.24674
# ## used   county level sampling rate: 51223351774.32022   # increase 0.3%
# ## used national level sampling rate: 37518194037.32022   # decrease 26.5%

# restored_actual_CBG_hourly_visitor_df = hourly_stop_sampling_rate_df.iloc[:, :hour_cnt + 1]

# print("Restored total visitor count (billion):")
# print(restored_actual_CBG_hourly_visitor_df.iloc[:, 1:hour_cnt + 1].sum().sum() / 1000000000)  # used CBG level sampling rate: 51028323628.24674

# print("Restored total visitor dataframe shape:",  restored_actual_CBG_hourly_visitor_df.shape)

In [ ]:
# restored_actual_CBG_hourly_visitor_df

In [ ]:
# print("Restored actuall CBG hourly visitor (billion):")
# print(restored_actual_CBG_hourly_visitor_df.iloc[:, 1:hour_cnt + 1].sum().sum()/ 1000000000)

In [ ]:
# restored_actual_target_hourly_visitor_df = pd.DataFrame()
# restored_actual_target_hourly_visitor_df['stop'] = restored_actual_CBG_hourly_visitor_df.iloc[:, 1:hour_cnt + 1].sum(axis=0) 
# print("Restored actuall target hourly visitor (billion):")
# print(restored_actual_target_hourly_visitor_df['stop'].sum() / 1000000000)
# restored_actual_target_hourly_visitor_df

In [ ]:
# # save the restored actual visitor count
# restored_actual_CBG_hourly_visitor_df.to_csv(restored_CBG_hourly_vistor_count_fname, index=False)
# restored_actual_target_hourly_visitor_df.to_csv(restored_target_hourly_vistor_count_fname, index=False)

## IPF

### Load data

In [ ]:
# restored_actual_target_hourly_visitor_df = pd.read_csv(restored_target_hourly_vistor_count_fname, dtype={'CBG': str})
# print("Restored target total visitor count (billion):")
# print(restored_actual_target_hourly_visitor_df['stop'].sum() / 1000000000) 

# restored_actual_target_hourly_visitor_df

Load the CBG stop distribution and compute the CBG aggregated restored visitors.

How to convert the CBG device distribution to stop distribution?

- option 1: convert the CBG device distribution to people distribution using device sampling rate, then using the people distribution as the stop distribution.
- option 2: compute the stop_per_device for each CBG, and then the stop for each home CBG, then aggregate the home CBG stops.

We use option 2 here.

In [ ]:
# CBG_device_stop_origin_distribution_df = pd.read_csv(CBG_device_stop_origin_distribution_fname, dtype={'origin':str})
# CBG_device_stop_origin_distribution_df

# print("origin observed stop count (billion):")
# print(CBG_device_stop_origin_distribution_df['stop'].sum() / 1000000000)

In [ ]:
# df = CBG_device_stop_origin_distribution_df.merge(people_per_device_df, left_on='origin', right_on='CBG_FIPS')
# df['restored_outgoer'] = df['stop'] * df['person_per_device']
# df['restored_outgoer'].sum() / 1000000000

In [ ]:



# visitor_per_stop = restored_actual_target_hourly_visitor_df['stop'].sum() / CBG_device_stop_origin_distribution_df['stop'].sum() 
# print("visitor_per_stop:", visitor_per_stop)

# restored_CBG_origin_outgoer_df = CBG_device_stop_origin_distribution_df[['origin', 'stop']].copy()
# restored_CBG_origin_outgoer_df['stop'] = CBG_device_stop_origin_distribution_df['stop'] * visitor_per_stop

# print("restored CBG origin outgoer count (billion):")
# restored_CBG_origin_outgoer_df['stop'].sum() / 1000000000

# restored_CBG_origin_outgoer_df

In [ ]:
# import numpy as np
# from ipfn import ipfn

# restored_outgoer_arr = restored_CBG_origin_outgoer_df['stop'].to_numpy()
# restored_actual_target_hourly_visitor_arr = restored_actual_target_hourly_visitor_df['stop'].to_numpy()

# aggregates = [restored_outgoer_arr, restored_actual_target_hourly_visitor_arr]
# dimensions = [[0], [1]]

# seed_arr = np.ones((len(restored_outgoer_arr), len(restored_actual_target_hourly_visitor_arr)))

# IPF = ipfn.ipfn(seed_arr, aggregates, dimensions, convergence_rate=1e-3)
# m = IPF.iteration()
# print(m.shape)

In [ ]:
# print(m.shape, restored_outgoer_arr.sum(), restored_actual_target_hourly_visitor_arr.sum())

# (m.sum(axis=1) - restored_outgoer_arr).sum(), (m.sum(axis=0) - restored_actual_target_hourly_visitor_arr).sum()

In [ ]:
# Save IPF results
# hourly_columns = [f'hour_{h+1}' for h in range(restored_actual_target_hourly_visitor_arr.shape[0])]
# IPF_df = pd.DataFrame(m, columns=hourly_columns)
# print("Total visitor (billion):", IPF_df.sum().sum() / 1000000000)

 
# IPF_df['CBG'] = restored_CBG_origin_outgoer_df['origin']
# IPF_df = IPF_df[['CBG'] + hourly_columns]

# IPF_df.to_csv(IPF_results_fname, index=False)
# IPF_df

## Compute hourly population map

## Load the CBG population

In [ ]:
if 'ACS_df'  not in locals():
    ACS_df = pd.read_csv(person_per_device_fname, dtype={"CBG_FIPS":str, "county_FIPS":str})
ACS_df

In [ ]:
if 'IPF_df'   not in locals():
    IPF_df = pd.read_csv(IPF_results_fname, dtype={'CBG':str}).set_index('CBG')

try:
    IPF_df.set_index('CBG', inplace=True)
except:
    pass

hour_cnt = IPF_df.shape[1]  
print(f"hour cnt: {hour_cnt}")

print("IPF total outbound (billion):", IPF_df.sum().sum()/ 10**9)

IPF_df

In [ ]:
inbound_hourly_df_fname

In [ ]:
if 'restored_actual_CBG_hourly_visitor_df'   not in locals():
    try:
        restored_actual_CBG_hourly_visitor_df = pd.read_csv(inbound_hourly_df_fname, encoding='utf-8', dtype={'CBG':str}).set_index('CBG')
    except:
        restored_actual_CBG_hourly_visitor_df = pd.read_csv(inbound_hourly_df_fname, encoding='utf-16', dtype={'CBG':str}).set_index('CBG')

print("Total restored visitor count (billion):", restored_actual_CBG_hourly_visitor_df.sum().sum() / 10**9)
restored_actual_CBG_hourly_visitor_df

Find the common CBGs between IPF and ACS CBGs

In [ ]:
CBG_device_stop_origin_distribution_df = pd.read_csv(CBG_device_stop_origin_distribution_fname, dtype={'origin':str})
CBG_device_stop_origin_distribution_df

In [ ]:
CBGs = set(IPF_df.index).intersection(set(ACS_df['CBG_FIPS']))
CBGs = CBGs.intersection(set(CBG_device_stop_origin_distribution_df['origin'])).intersection(set(restored_actual_CBG_hourly_visitor_df.index))
CBGs = list(CBGs)
print("Found CBG count:", len(CBGs))

In [ ]:
filtered_ACS_df = ACS_df.set_index("CBG_FIPS").loc[CBGs].sort_index()

filtered_outgoer_df = IPF_df.loc[CBGs].sort_index()
filtered_visitor_df = restored_actual_CBG_hourly_visitor_df.loc[CBGs].iloc[:, :hour_cnt]

print("Filtered restored outgoer count (billion):", filtered_outgoer_df.sum().sum() / 1000000000)

print("Filtered restored visitor count (billion):", filtered_visitor_df.sum().sum() / 1000000000)

filtered_outgoer_df
filtered_visitor_df

In [ ]:
hourly_population_base_df = filtered_outgoer_df.copy()

for c in tqdm(hourly_population_base_df.columns):
    hourly_population_base_df[c] = filtered_ACS_df['totalpopulation'].to_list()

hourly_population_base_df

In [ ]:
hourly_population_map_df = hourly_population_base_df - filtered_outgoer_df + filtered_visitor_df 

print("Saving hourly map at: ", CBG_hourly_population_fname)
hourly_population_map_df.to_csv(CBG_hourly_population_fname)


hourly_population_map_df

In [ ]:
filtered_visitor_df#.sum(axis=0)

In [ ]:
filtered_outgoer_df#.sum(axis=0)

# hourly_population_df = filtered_IPF_df.copy()
# for c in hourly_population_df.columns:
#     hourly_population_df[c] = filtered_population_df['totalpopulation'].to_list()

In [ ]:
hourly_population_map_df.sum(axis=0)

In [ ]:
# n = 10
# hourly_population_map_df = hourly_population_base_df.iloc[:n] - filtered_outgoer_df.iloc[:n] + filtered_visitor_df.iloc[:n] 

## Show hourly population map

In [ ]:
# Load data
hourly_population_map_df = pd.read_csv(CBG_hourly_population_fname, dtype={'CBG':str})
hourly_population_map_df

In [ ]:
# hourly_population_map_df

In [ ]:
ACS_df = pd.read_csv(person_per_device_fname, dtype={"CBG_FIPS":str, "county_FIPS":str})
ACS_df['county_FIPS'] = ACS_df['county_FIPS'].astype(str).str.zfill(5)
ACS_df

In [ ]:
import geopandas as gpd

def get_county_FIPS_dict():
    county_gdf = gpd.read_file(r'https://github.com/gladcolor/spatial_data/raw/master/cb_2019_us_county_20m.zip')
    state_gdf = gpd.read_file(r'https://github.com/gladcolor/spatial_data/raw/master/cb_2019_us_state_20m.zip')
    county_gdf = county_gdf.merge(state_gdf[['STATEFP', 'STUSPS', 'NAME']], left_on='STATEFP', right_on='STATEFP', suffixes=['_county', '_state'])
    county_gdf['county_state'] = county_gdf['NAME_county'] + ', ' + county_gdf['STUSPS']
    county_gdf['GEOID'] = county_gdf['GEOID'].astype(str).str.zfill(5)
    county_FIPS_dict = county_gdf.set_index(['county_state'])['GEOID'].to_dict()
    return county_FIPS_dict

county_FIPS_dict = get_county_FIPS_dict()
# county_FIPS_dict

In [ ]:
# CBG_2019_fname = r"D:\OneDrive_PSU\OneDrive - The Pennsylvania State University\Research_doc\Wild_fire\blockgoups2019.zip"
CBG_2019_fname = r"E:\OneDrive_PSU\OneDrive - The Pennsylvania State University\Research_doc\Wild_fire\blockgoups2019.zip"  # 2018 

CBG2019_gdf = gpd.read_file(CBG_2019_fname)
CBG2019_gdf['county_FIPS'] = CBG2019_gdf['GEOID'].astype(str).str.zfill(12).str[:5]
CBG2019_gdf['CBG'] = CBG2019_gdf['GEOID'].astype(str).str.zfill(12) 


In [ ]:
CBG2019_gdf

In [ ]:
 # hourly_population_map_df

In [ ]:
hour_cnt = hourly_population_map_df.shape[1] - 2
print("hour_cnt:", hour_cnt)

In [ ]:
hourly_population_map_df.iloc[:, 1:hour_cnt]

In [ ]:
import warnings

# Suppress all future warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

cmap = 'viridis'
# day = 18

target_county = 'New York, NY'
target_county = 'Centre, PA'
target_county = 'Richland, SC'
target_county = 'Lexington, SC'
target_county = 'Union, NJ'
target_county = 'Fairfax, VA'
target_county = 'Los Angeles, CA'
target_county = 'Cook, IL'
target_county = 'Harris, TX'
target_county = 'Maricopa, AZ'
target_county = 'San Diego, CA'
target_county = 'Orange, CA'
target_county = 'Miami-Dade, FL'
target_county = 'Dallas, TX'
target_county = 'King, WA'
target_county = 'Riverside, CA'
target_county = 'Hudson, NJ'
target_county = 'Essex, NJ'
target_county = 'Berkshire, MA'


def draw_county_map(target_county):
    target_county = 'New York, NY'
    county_FIPS = county_FIPS_dict[target_county]
    
    target_counties = [county_FIPS] # ['36061']   # 06039
    hourly_population_map_df['county_FIPS'] = hourly_population_map_df.reset_index()['CBG'].astype(str).str[:5].to_list()
    target_hourly_pop_df = hourly_population_map_df[hourly_population_map_df['county_FIPS'].isin(target_counties)]
    
    target_pop_df = ACS_df[ACS_df['county_FIPS'].isin(target_counties)]
    
    
    target_blockgroup_gdf = CBG2019_gdf[CBG2019_gdf['county_FIPS'].isin(target_counties)]
    target_blockgroup_gdf
    
    row_cnt = 2
    col_cnt = 13
    # vmax = max(target_hourly_pop_df.iloc[:, :hour_cnt].max(axis=1).max(), target_pop_df['totalpopulation'].max())
    # vmax = target_hourly_pop_df.iloc[:, :hour_cnt ].mean(axis=1).mean() + target_hourly_pop_df.iloc[:, :hour_cnt].mean(axis=1).std() * 3
    print("target_hourly_pop_df:\n", target_hourly_pop_df.iloc[:, 1: hour_cnt + 1].mean(axis=1).mean())
    vmax = target_hourly_pop_df.iloc[:, 1: hour_cnt + 1].mean(axis=1).mean() + target_hourly_pop_df.iloc[:, 1:hour_cnt + 1].mean(axis=1).std() * 3
    
    print("vmax: ", vmax)
    fig, axs = plt.subplots(figsize=(40, 10), nrows=row_cnt, ncols=col_cnt)
    
    # draw hourly map
    used_ax_cnt = 0
    hour_24_pop = []
    for row in range(row_cnt):
        for col in range(1, col_cnt, 1):
            # if used_ax_cnt == 24:
            #     continue
            merged_gdf = target_blockgroup_gdf.merge(target_hourly_pop_df, left_on='CBG', right_on='CBG')
            hour = used_ax_cnt
            ax=axs[row, col]
            ax.axis('off')
            ax.set_title(f"{hour:02}:00:00")
            hour_column = f'hour_{hour + 1 + 24*(day - 1)}'
            merged_gdf.plot(column=hour_column, ax=ax, vmax=vmax, cmap=cmap)  # , vmax=vmax ,  legend=True
            # merged_gdf.plot(column=hour_column, ax=ax, cmap=cmap)
            hour_24_pop.append(target_hourly_pop_df[hour_column].sum())
            used_ax_cnt += 1
    
    ax=axs[0, 0]
    ax.axis('off')
    ax.set_title(f"ACS 2019 Population")
    merged_gdf = target_blockgroup_gdf.merge(target_pop_df, left_on='CBG', right_on='CBG_FIPS')
    # merged_gdf.plot(column='totalpopulation', ax=ax,  cmap=cmap) # , vmax=vmax ,  legend=True
    merged_gdf.plot(column='totalpopulation', ax=ax, vmax=vmax, cmap=cmap)
    
    # draw county-level population
    ax=axs[1, 0]
    # ax.axis('off')
    ax.set_title('County hourly population')
    ax.set_ylabel('Population')
    ax.set_xlabel('Hour')
    ax.plot(hour_24_pop, label='hourly population')
    ax.axhline(merged_gdf['totalpopulation'].sum(), color='red', label='ACS population')
    ax.legend()
    
     
    fig.suptitle(f"Hourly population of county {target_county}: {year}-{month:02}-{day:02}", fontsize=24, y=1)
    
    
    # Add colorbar axes at the bottom and align it with the left and right of the subplots
    pos1 = axs[0, 0].get_position() # get the original position for first axis
    pos2 = axs[-1, -1].get_position()
    cax = fig.add_axes([pos1.x0 * 1.01, 0, (pos2.x1 - pos1.x0) * 0.99, 0.008])    # [left, bottom, width, height]  # bottom
    value_max =vmax
    value_min = 0
    
    sm = plt.cm.ScalarMappable(cmap=cmap, norm=plt.Normalize(vmin=value_min, vmax=value_max))
    sm._A = []
    # plt.colorbar(sm, cax=cax, label="Bias")
    cbar = plt.colorbar(sm, cax=cax, orientation='horizontal')
    
    for t in cbar.ax.get_xticklabels():
        t.set_fontsize(16)
    
    cbar.set_label('Population', labelpad=-80,   rotation=0, fontsize=24, loc='center')
    
    png_name = os.path.join(save_dir, f"{target_county.replace(', ', '_')}_hourly_population_{year}{month:02}{day:02}_v3.png")
    
    print("PNG name:", png_name)
    
    plt.savefig(png_name, dpi=300, bbox_inches='tight', pad_inches=0.2)
    plt.close()

for target_county, FIPS in tqdm(county_FIPS_dict.items()):
    print(target_county)
    for day in range(1, 32):
        draw_county_map(target_county)
    break

print("Done")

In [ ]:
target_hourly_pop_df
merged_gdf
hourly_population_map_df
target_blockgroup_gdf
target_county

In [ ]:
STOP

# Extract data for South Carolina

In [ ]:
state_FIPS = '45'
year = '2023'
month = 6

houry_population_fname = os.path.join(save_dir, 'for_South_Carolina', f'state_FIPS_{state_FIPS:02}_hourly_population_{year}{month:02}.csv')
print('houry_population_fname:', houry_population_fname)

inbound_fname = os.path.join(save_dir, 'for_South_Carolina', f'state_FIPS_{state_FIPS:02}_hourly_inbound_{year}{month:02}.csv')
print('inbound_fname:', inbound_fname)

outbound_fname = os.path.join(save_dir, 'for_South_Carolina', f'state_FIPS_{state_FIPS:02}_hourly_outbound_{year}{month:02}.csv')
print('inbound_fname:', outbound_fname)

device_distribution_fname = os.path.join(save_dir, 'for_South_Carolina', f'state_FIPS_{state_FIPS:02}_monthly_device_distribution_{year}{month:02}.csv')
print('device_distribution_fname:', device_distribution_fname)

device_sampling_rate_fname = os.path.join(save_dir, 'for_South_Carolina', f'state_FIPS_{state_FIPS:02}_monthly_device_sampling_rate_{year}{month:02}.csv')
print('device_sampling_rate_fname:', device_sampling_rate_fname)

In [ ]:
# Outbound
IPF_df = pd.read_csv(IPF_results_fname, dtype={'CBG':str}).set_index('CBG')

hour_cnt = IPF_df.shape[1]  
print(f"hour cnt: {hour_cnt}")

print("IPF total outbound (billion):", IPF_df.iloc[:, :].sum().sum()/ 10**9)

IPF_df

In [ ]:
# outbound_fname
extracted_IPF_df = IPF_df[IPF_df.index.str.startswith(state_FIPS)]
extracted_IPF_df.to_csv(outbound_fname, index=True)
extracted_IPF_df

In [ ]:
# Inbound
try:
    restored_actual_CBG_hourly_visitor_df = pd.read_csv(inbound_hourly_df_fname, encoding='utf-8', dtype={'CBG':str}).set_index('CBG')
except:
    restored_actual_CBG_hourly_visitor_df = pd.read_csv(inbound_hourly_df_fname, encoding='utf-16', dtype={'CBG':str}).set_index('CBG')

restored_actual_CBG_hourly_visitor_df = restored_actual_CBG_hourly_visitor_df.iloc[:, :hour_cnt]
print("Total restored visitor count (billion):", restored_actual_CBG_hourly_visitor_df.sum().sum() / 10**9)
restored_actual_CBG_hourly_visitor_df

In [ ]:
# inbound_fname
extracted_inboud_df = restored_actual_CBG_hourly_visitor_df[restored_actual_CBG_hourly_visitor_df.index.str.startswith(state_FIPS)]
extracted_inboud_df.to_csv(inbound_fname, index=True)
extracted_inboud_df

In [ ]:
# Load the hourly map
hourly_population_map_df = pd.read_csv(CBG_hourly_population_fname, dtype={'CBG':str}).set_index('CBG')
hourly_population_map_df

In [ ]:
# houry_population_fname
houry_population_df = hourly_population_map_df[hourly_population_map_df.index.str.startswith(state_FIPS)]
houry_population_df.to_csv(houry_population_fname, index=True)
houry_population_df

In [ ]:
# device distribution
device_home_areas_df = pd.read_csv(device_home_areas_fname, dtype={"origin":str, "destination":str})
device_home_areas_df


In [ ]:
# device_distribution_fname
extracted_device_distribution_df = device_home_areas_df[device_home_areas_df['destination'].str.startswith(state_FIPS)]
extracted_device_distribution_df = extracted_device_distribution_df[extracted_device_distribution_df['origin'].str.startswith(state_FIPS)]

extracted_device_distribution_df.to_csv(device_distribution_fname, index=False)
extracted_device_distribution_df

In [ ]:
# Load device sampling rate
print("Loading people_per_device_df from:", person_per_device_fname)
person_per_device_df = pd.read_csv(person_per_device_fname, dtype={"CBG_FIPS": str})

print("Total totalpopulation (billion):", person_per_device_df['totalpopulation'].sum() / 10**9)
person_per_device_df

In [ ]:
# device_sampling_rate_fname
extracted_person_per_device_df = person_per_device_df[person_per_device_df['CBG_FIPS'].str.startswith(state_FIPS)]
extracted_person_per_device_df.to_csv(device_sampling_rate_fname, index=False)
extracted_person_per_device_df